File: exercises/dotnet/InventoryService/Domain/Identifiers.cs
using SharedObservability;

namespace InventoryService.Domain;

/// <summary>
/// Domain identifiers for the Inventory bounded context.
/// Each enum value maps to a structured-logging / baggage key.
/// </summary>
public enum InventoryContextKey
{
    ORDER_ID,
    CUSTOMER_ID,
    RESERVATION_ID,
    PRODUCT_CODE
}

public static class InventoryContextKeyExtensions
{
    private static readonly Dictionary<InventoryContextKey, string> Keys = new()
    {
        [InventoryContextKey.ORDER_ID] = "order.id",
        [InventoryContextKey.CUSTOMER_ID] = "customer.id",
        [InventoryContextKey.RESERVATION_ID] = "reservation.id",
        [InventoryContextKey.PRODUCT_CODE] = "product.code",
    };

    /// <summary>Returns the dotted key name used in baggage and structured logs.</summary>
    public static string KeyName(this InventoryContextKey key) => Keys[key];

    /// <summary>Creates an <see cref="IDomainIdentifier"/> binding this key to a value.</summary>
    public static IDomainIdentifier Of(this InventoryContextKey key, string value) =>
        new SimpleIdentifier(Keys[key], value);

    private record SimpleIdentifier(string Key, string Value) : IDomainIdentifier;
}

File: exercises/python/inventory_service/domain/identifiers.py
"""Domain identifiers for the Inventory bounded context.

Each enum value carries a dotted key string that matches the wire contract
shared by convention (not shared kernel) across bounded contexts —
e.g. ``"order.id"`` is the same string Order uses.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from shared_observability import DomainIdentifier


# ---------------------------------------------------------------------------
# Private helper — satisfies the DomainIdentifier protocol structurally
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _SimpleIdentifier:
    _key: str
    _value: str

    def key(self) -> str:
        return self._key

    def value(self) -> str:
        return self._value


# ---------------------------------------------------------------------------
# Context keys for structured logging / OTel attributes
# ---------------------------------------------------------------------------


class InventoryContextKey(Enum):
    """Inventory-specific domain identifiers.

    Usage::

        with DomainContext(
            InventoryContextKey.ORDER_ID.of("ord_abc"),
            InventoryContextKey.RESERVATION_ID.of("res_xyz"),
        ):
            logger.info("reserved")
    """

    ORDER_ID = "order.id"
    CUSTOMER_ID = "customer.id"
    RESERVATION_ID = "reservation.id"
    PRODUCT_CODE = "product.code"

    def of(self, value: str) -> DomainIdentifier:
        """Create a :class:`DomainIdentifier` for this key with *value*."""
        return _SimpleIdentifier(_key=self.value, _value=value)
